segmentation/utils/spatialorgmarkerexpression_utils.py

```diff
@@ -0,0 +1,265 @@
+            (muhat[j, k], sigmahat[j, k]) = scipy.stats.norm.fit(tmp)
+            # calculate z score based on distribution
+            z[j, k] = (close_num[j, k] - muhat[j, k]) / sigmahat[j, k]
+            # calculate both positive and negative enrichment p values
+            p[j, k, 0] = (1 + (np.sum(tmp >= close_num[j, k]))) / (bootstrap_num + 1)
+            p[j, k, 1] = (1 + (np.sum(tmp <= close_num[j, k]))) / (bootstrap_num + 1)
+
+    # get fdh_br adjusted p values
+    p_summary = np.zeros_like(p[:, :, 0])
+    for j in range(0, marker_num):
+        for k in range(0, marker_num):
+            # use negative enrichment p values if the z score is negative, and vice versa
+            if z[j, k] > 0:
+                p_summary[j, k] = p[j, k, 0]
+            else:
+                p_summary[j, k] = p[j, k, 1]
+    (h, adj_p, aS, aB) = statsmodels.stats.multitest.multipletests(
+        p_summary, alpha=.05)
+
+    return z, muhat, sigmahat, p, h, adj_p
+
+
+def calculate_channel_spatial_enrichment(dist_matrix, marker_thresholds, cell_array,
+                                         excluded_colnames=None, points=None,
+                                         patient_idx=30, dist_lim=100, cell_label_idx=24, bootstrap_num=100):
+    """Spatial enrichment analysis to find significant interactions between cells expressing different markers.
+    Uses bootstrapping to permute cell labels randomly.
+
+    Args:
+        dist_matrix: cells x cells matrix with the euclidian
+            distance between centers of corresponding cells
+        marker_thresholds: threshold values for positive marker expression
+        cell_array: data including points, cell labels, and
+            cell expression matrix for all markers
+        excluded_colnames: all column names that are not markers
+        points: patient labels to include in analysis
+        patient_idx: columns with patient labels
+        dist_lim: cell proximity threshold
+        cell_label_idx: column with cell labels
+        bootstrap_num: number of permutations for bootstrap
+
+    Returns:
+        values: a list with each element consisting of a tuple of
+            closenum and closenumrand for each point included in the analysis
+        stats: a list with each element consisting of a tuple of
+            z, muhat, sigmahat, p, h, adj_p, and marker_titles for each point in the analysis"""
+
+    # Setup input and parameters
+    if points is None:
+        points = list(set(cell_array.iloc[:, patient_idx]))
+    values = []
+    stats = []
+
+    if excluded_colnames is None:
+        excluded_colnames = ["cell_size", "Background", "HH3",
+                             "summed_channel", "label", "area",
+                             "eccentricity", "major_axis_length", "minor_axis_length",
+                             "perimeter", "fov"]
+
+    # Error Checking
+    if not np.isin(excluded_colnames, cell_array.columns).all():
```

Comment:
This doesn't actually raise an error, just prints a message. Check out the error checking in segmentation_utils where we raise a ValueError to prevent the function from continuing